Error : "unable to start CRD watcher"

Running a latest master build on a highly restricted cluster I get the following error:

```
build/octant -v --namespace <NAMESPACE> --disable-cluster-overview
2020-01-23T15:49:55.677Z        DEBUG   commands/dash.go:67     disable-open-browser: %!s(bool=false)
2020-01-23T15:49:55.677Z        DEBUG   dash/dash.go:70 Loading configuration: /home/<USERID>/.kube/config
2020-01-23T15:49:55.680Z        DEBUG   cluster/cluster.go:265  initializing REST client configuration  {"client-qps": 200, "client-burst": 400}
2020-01-23T15:49:55.680Z        DEBUG   cluster/cluster.go:108  created temp directory  {"component": "cluster client", "dir": "/home/<USERID>/Library/cache/octant494091269"}
2020-01-23T15:49:55.680Z        DEBUG   dash/dash.go:98 initial namespace for dashboard is <NAMESPACE>
2020-01-23T15:49:56.039Z        ERROR   commands/dash.go:109    dashboard failed        {"err": "unable to start CRD watcher"}
github.com/vmware-tanzu/octant/internal/commands.newOctantCmd.func1.2
        /home/<USERID>/Library/octant/internal/commands/dash.go:109
```

Is this an error in the build itself or as a result of not being able to list CRDs from an RBAC perspective? Running on RHEL 7 Linux.

```
git log -n 1 --oneline
81e55320 (HEAD -> issue-562, origin/issue-562) fix AccessError type assertion in CRD watcher
```
```
build/octant version
Version:  (dev-version)
Git commit:  81e55320
Built:  2020-01-24T20:50:38Z
```
```
env OCTANT_LISTENER_ADDR=127.0.0.1:8900 build/octant -v --namespace <NAMESPACE> --disable-cluster-overview
2020-01-24T20:55:21.368Z        DEBUG   commands/dash.go:67     disable-open-browser: %!s(bool=false)
2020-01-24T20:55:21.368Z        DEBUG   dash/dash.go:69 Loading configuration: /home/<USERID>/.kube/config
2020-01-24T20:55:21.370Z        DEBUG   cluster/cluster.go:265  initializing REST client configuration  {"client-qps": 200, "client-burst": 400}
2020-01-24T20:55:21.371Z        DEBUG   cluster/cluster.go:108  created temp directory  {"component": "cluster client", "dir": "/home/<USERID>/Library/cache/octant493982554"}
2020-01-24T20:55:21.371Z        DEBUG   dash/dash.go:97 initial namespace for dashboard is <NAMESPACE>
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x137e5c5]

goroutine 11 [running]:
github.com/vmware-tanzu/octant/internal/errors.(*AccessError).Error(0xc00013e090, 0xc000236d30, 0xc00074da08)
        /home/<USERID>/Library/octant_test/octant/internal/errors/access_error.go:47 +0x45
github.com/vmware-tanzu/octant/internal/log.(*sugaredLogWrapper).WithErr(0xc0001300b8, 0x2546f80, 0xc00013e090, 0xc000672100, 0x1)
        /home/<USERID>/Library/octant_test/octant/internal/log/logger.go:45 +0x35
github.com/vmware-tanzu/octant/internal/describer.(*DefaultCRDWatcher).Watch(0xc00037eec0, 0x2588780, 0xc0005c7a70, 0xc000231480, 0x25a76c0)
        /home/<USERID>/Library/octant_test/octant/internal/describer/crd_watcher.go:142 +0x3a5
github.com/vmware-tanzu/octant/pkg/dash.Run(0x25886c0, 0xc000566880, 0x259d520, 0xc0001300b8, 0xc000048230, 0x100, 0xc000041440, 0x1a, 0x7ffcd457ce32, 0x12, ...)
        /home/<USERID>/Library/octant_test/octant/pkg/dash/dash.go:114 +0x746
github.com/vmware-tanzu/octant/internal/commands.newOctantCmd.func1.2(0x19b49d9, 0xd, 0x25886c0, 0xc000566880, 0x259d520, 0xc0001300b8, 0xc000048230, 0xc0000481c0)
        /home/<USERID>/Library/octant_test/octant/internal/commands/dash.go:108 +0x5e3
created by github.com/vmware-tanzu/octant/internal/commands.newOctantCmd.func1
        /home/<USERID>/Library/octant_test/octant/internal/commands/dash.go:73 +0x518
```

--- a/changelogs/unreleased/562-wwitzel3
+++ b/changelogs/unreleased/562-wwitzel3
@@ -1,0 +1,1 @@
+Fix error type assertion for AccessError.

--- a/internal/describer/crd_watcher.go
+++ b/internal/describer/crd_watcher.go
@@ -7,11 +7,11 @@ package describer
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"sync"
 
 	"github.com/google/uuid"
-	"github.com/pkg/errors"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	kcache "k8s.io/client-go/tools/cache"
 
@@ -31,7 +31,6 @@ type DefaultCRDWatcher struct {
 
 	watchConfigs map[string]*config.CRDWatchConfig
 
-	accessError sync.Once
 	mu          sync.Mutex
 }
 
@@ -135,16 +134,16 @@ func (cw *DefaultCRDWatcher) Watch(ctx context.Context) error {
 
 	err := cw.objectStore.Watch(ctx, crdKey, handler)
 	if err != nil {
-		aErr, ok := err.(*internalErr.AccessError)
-		if ok {
-			found := cw.errorStore.Add(aErr)
+		var e *internalErr.AccessError
+		if errors.As(err, &e) {
+			found := cw.errorStore.Add(e)
 			// Log if we have not seen this access error before.
 			if !found {
-				logger.WithErr(aErr).Errorf("access denied")
+				logger.WithErr(e).Errorf("access denied")
 			}
 			return nil
 		}
-		return errors.WithMessage(err, "crd watcher has failed")
+		return fmt.Errorf("crd watcher has failed: %w", err)
 	}
 
 	return nil

--- a/internal/errors/access_error.go
+++ b/internal/errors/access_error.go
@@ -44,7 +44,11 @@ func (o *AccessError) Timestamp() time.Time {
 
 // Error returns an error string.
 func (o *AccessError) Error() string {
-	return fmt.Sprintf("%s: %s: %s", o.verb, o.key, o.err)
+	e := fmt.Sprintf("%s: %s", o.verb, o.key)
+	if o.err != nil {
+		e = fmt.Sprintf("%s (error: %s)", e, o.err.Error())
+	}
+	return e
 }
 
 // Key returns the key for the error.

--- a/internal/errors/access_error_test.go
+++ b/internal/errors/access_error_test.go
@@ -6,6 +6,7 @@
 package errors
 
 import (
+	goerrors "errors"
 	"fmt"
 	"testing"
 
@@ -25,8 +26,36 @@ func TestNewAccessError(t *testing.T) {
 	intErr := NewAccessError(key, verb, err)
 	assert.Equal(t, key, intErr.Key())
 	assert.Equal(t, verb, intErr.Verb())
-	assert.Equal(t, fmt.Sprintf("%s: %s: %s", verb, key, err), intErr.Error())
+	assert.Equal(t, fmt.Sprintf("%s: %s (error: %s)", verb, key, err.Error()), intErr.Error())
 	assert.EqualError(t, err, "access denied")
 	assert.NotEmpty(t, intErr.Timestamp())
 	assert.NotZero(t, intErr.ID())
+}
+
+func TestFormattedAccessError(t *testing.T) {
+	key := store.Key{
+		Namespace:  "default",
+		APIVersion: "v1",
+		Kind:       "Pod",
+	}
+	verb := "watch"
+	err := fmt.Errorf("access denied")
+
+	intErr := NewAccessError(key, verb, err)
+	newErr := fmt.Errorf("%w", intErr)
+
+	var e *AccessError
+	assert.True(t, goerrors.As(newErr, &e))
+}
+
+func TestNilErrAccessError(t *testing.T) {
+	key := store.Key{
+		Namespace:  "default",
+		APIVersion: "v1",
+		Kind:       "Pod",
+	}
+	verb := "watch"
+
+	intErr := NewAccessError(key, verb, nil)
+	assert.Equal(t, fmt.Sprintf("%s: %s", verb, key), intErr.Error())
 }

--- a/internal/objectstore/access.go
+++ b/internal/objectstore/access.go
@@ -10,7 +10,6 @@ import (
 	"fmt"
 	"sync"
 
-	"github.com/pkg/errors"
 	"go.opencensus.io/trace"
 	authorizationv1 "k8s.io/api/authorization/v1"
 
@@ -133,7 +132,7 @@ func (r *resourceAccess) HasAccess(ctx context.Context, key store.Key, verb stri
 		span.Annotate([]trace.Attribute{}, "fetch access start")
 		val, err := r.fetchAccess(aKey, verb)
 		if err != nil {
-			return errors.Wrapf(err, "fetch access: %+v", aKey)
+			return fmt.Errorf("fetch access: %+v: %w", aKey, err)
 		}
 
 		r.cache.set(aKey, val)
@@ -155,7 +154,7 @@ func (r *resourceAccess) keyToAccessKey(key store.Key, verb string) (AccessKey, 
 	gvk := key.GroupVersionKind()
 
 	if gvk.GroupKind().Empty() {
-		return AccessKey{}, errors.Errorf("unable to check access for key %s", key.String())
+		return AccessKey{}, fmt.Errorf("unable to check access for key %s", key.String())
 	}
 
 	gvr, err := r.client.Resource(gvk.GroupKind())
@@ -178,7 +177,7 @@ func (r *resourceAccess) fetchAccess(key AccessKey, verb string) (bool, error) {
 
 	k8sClient, err := r.client.KubernetesClient()
 	if err != nil {
-		return false, errors.Wrap(err, "client kubernetes")
+		return false, fmt.Errorf("client kubernetes: %w", err)
 	}
 
 	authClient := k8sClient.AuthorizationV1()
@@ -195,7 +194,7 @@ func (r *resourceAccess) fetchAccess(key AccessKey, verb string) (bool, error) {
 
 	review, err := authClient.SelfSubjectAccessReviews().Create(sar)
 	if err != nil {
-		return false, errors.Wrap(err, "client auth")
+		return false, fmt.Errorf("client auth: %w", err)
 	}
 	return review.Status.Allowed, nil
 }

--- a/internal/objectstore/dynamic_cache.go
+++ b/internal/objectstore/dynamic_cache.go
@@ -7,11 +7,11 @@ package objectstore
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"sync"
 	"time"
 
-	"github.com/pkg/errors"
 	"go.opencensus.io/trace"
 	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/api/meta"
@@ -113,7 +113,7 @@ func NewDynamicCache(ctx context.Context, client cluster.ClientInterface, option
 
 	factory, err := c.initFactoryFunc(context.Background(), client, "")
 	if err != nil {
-		return nil, errors.Wrap(err, "initialize dynamic shared informer factory")
+		return nil, fmt.Errorf("initialize dynamic shared informer factory: %w", err)
 	}
 
 	c.factories.set("", factory)
@@ -204,7 +204,7 @@ func (dc *DynamicCache) listFromInformer(ctx context.Context, key store.Key) (*u
 
 	informer, hasSynced, err := dc.currentInformer(ctx, key)
 	if err != nil {
-		return nil, false, errors.Wrapf(err, "retrieving informer for %+v", key)
+		return nil, false, fmt.Errorf("retrieving informer for %+v: %w", key, err)
 	}
 
 	if !hasSynced {
@@ -226,7 +226,7 @@ func (dc *DynamicCache) listFromInformer(ctx context.Context, key store.Key) (*u
 
 	objects, err := l.List(selector)
 	if err != nil {
-		return nil, false, errors.Wrapf(err, "listing %v", key)
+		return nil, false, fmt.Errorf("listing %v: %w", key, err)
 	}
 
 	list := &unstructured.UnstructuredList{}
@@ -306,7 +306,7 @@ func (dc *DynamicCache) getFromInformer(ctx context.Context, key store.Key) (*un
 
 	informer, hasSynced, err := dc.currentInformer(ctx, key)
 	if err != nil {
-		return nil, errors.Wrapf(err, "retrieving informer for %v", key)
+		return nil, fmt.Errorf("retrieving informer for %v: %w", key, err)
 	}
 
 	if !hasSynced {
@@ -356,7 +356,7 @@ func (dc *DynamicCache) Watch(ctx context.Context, key store.Key, handler kcache
 
 	informer, _, err := dc.currentInformer(ctx, key)
 	if err != nil {
-		return errors.Wrapf(err, "retrieving informer for %s", key)
+		return fmt.Errorf("retrieving informer for %s: %w", key, err)
 	}
 
 	informer.Informer().AddEventHandler(handler)
@@ -447,7 +447,7 @@ func (dc *DynamicCache) Update(ctx context.Context, key store.Key, updater func(
 		}
 
 		if !found {
-			return errors.Errorf("object not found")
+			return errors.New("object not found")
 		}
 
 		gvk := object.GroupVersionKind()
@@ -463,7 +463,7 @@ func (dc *DynamicCache) Update(ctx context.Context, key store.Key, updater func(
 		}
 
 		if err := updater(object); err != nil {
-			return errors.Wrap(err, "unable to update object")
+			return fmt.Errorf("unable to update object: %w", err)
 		}
 
 		client := dynamicClient.Resource(gvr).Namespace(object.GetNamespace())

